### ISSUE TITLE
jgrapht.alg.flow.PushRelabelMFImpl.getMaximumFlow has Ω(N^5) complexity

### ISSUE THREAD

**Issue by MrVPlusOne** [jgrapht/jgrapht#461]
Hi,

I found an input pattern that can trigger Ω(N^5) running time behavior of `jgrapht.alg.flow.PushRelabelMFImpl.getMaximumFlow`, the input is constructed and tested using the following code:
```java
import org.jgrapht.alg.flow.PushRelabelMFImpl;
import org.jgrapht.graph.DefaultWeightedEdge;
import org.jgrapht.graph.DirectedWeightedMultigraph;

public class BugReproduce {

    public static void main(String[] args) {
        testPushRelabelMFImpl(200);
    }

    public static void testPushRelabelMFImpl(int graphSize){
        int outWeight = 3;
        int inWeight = 4;

        DirectedWeightedMultigraph<Integer, DefaultWeightedEdge> g = new DirectedWeightedMultigraph<>(DefaultWeightedEdge.class);
        for (int i = 0; i<graphSize; i++){
            g.addVertex(i);
        }

        for(int i = 1; i < graphSize; i++) {
            if(i%2 == 0){
                DefaultWeightedEdge eOut = g.addEdge(0, i);
                g.setEdgeWeight(eOut, outWeight);

                DefaultWeightedEdge eInt = g.addEdge(i, 0);
                g.setEdgeWeight(eInt, inWeight);
            }else{
                DefaultWeightedEdge e = g.addEdge(0,i);
                g.setEdgeWeight(e, 0);
            }
        }

        System.out.println("Graph constructed... Size = " + graphSize);

        long startTime = System.nanoTime();
        PushRelabelMFImpl<Integer, DefaultWeightedEdge> algorithm = new PushRelabelMFImpl<>(g);
        algorithm.getMaximumFlow(2,1);

        double time = (System.nanoTime() - startTime)/1e9;

        System.out.println("Finished. Time used: " + time + "s.");
    }
}
```
On my machine, when `graphSize=200`, it took 11 secods to run this code, and when `graphSize=300`, it took 84s.

I have also tried to fit the performance-graphSize relationship, the result is a 5th-order curve:

<img width="707" alt="fit" src="https://user-images.githubusercontent.com/11240361/34233214-ac9484ee-e5a9-11e7-822f-13d8733ff5c8.png">

**Comment by jkinable** [comment]
Thanks for the detailed plot. Technically the implementation should be bounded by O(n^3), so this seems indeed a little off. However, in order to figure out what goes wrong, we need to do some more analysis (profiling) to figure out where the majority of computation time is spent.
One potential cause could be that you use a multigraph. Originally, the algorithm was developed for simple graphs. We would need to check what happens with multi graphs. Notice the comment in `PushRelabelImpl`: "Note: even though the algorithm accepts any kind of graph, currently only Simple directed and undirected graphs are supported (and tested!).

**Comment by jkinable** [comment]
@MrVPlusOne a fix has been submitted for the issue you posted (PR #535) . Could you test whether this fixes your problem?

**Comment by MrVPlusOne** [comment]
Thanks for the fix! We reran our tool on the new implementation and found that the new worst-case complexity to be N^3. It looks like the problem has been fixed.

**Comment by jkinable** [comment]
@MrVPlusOne That sounds about right: the push relabel alg has a runtime complexity of O(V^2*E). Thanks for the verification!

**Comment by MrVPlusOne** [comment]
Yeah, that sounds right. In the pattern we found, E grows proportionally with V, so N^3 falls within the expected complexity :)

**Comment by jkinable** [comment]
This issue has been fixed in PR #461. Thanks for reporting!

**Comment by jkinable** [comment]
I'm reopening this issue. When I run our performance tests on the max flow algorithms, there still seems to be an issue with the Push-Relabel implementation:
```
MaximumFlowAlgorithmPerformanceTest.DinicMaximumFlowRandomGraphBenchmark.run        avgt    5   134071652.771 ±  43998717.052  ns/op
MaximumFlowAlgorithmPerformanceTest.EdmondsKarpMaximumFlowRandomGraphBenchmark.run  avgt    5   158869711.971 ±  11239887.264  ns/op
MaximumFlowAlgorithmPerformanceTest.PushRelabelMaximumFlowRandomGraphBenchmark.run  avgt    5  7029793771.800 ± 294215326.892  ns/op
```
The PushRelabel implementation is about 70 times slower than Dinic/EdmondsKarp. This is much slower than it is supposed to be. See for instance this comparison:
Friis, Olesen,.An experimental comparison of max flow algorithms, http://www.cs.au.dk/~gerth/advising/thesis/jakob_mark_friis_steffen_beier_olesen.pdf
PushRelabel has a O(V^2E) time complexity, which is asymptotically more efficient than the O(VE^2) of Edmonds–Karp algorithm

**Comment by AlexandruValeanu** [comment]
@jkinable I am currently working on a different implementation for Push-Relabel. I'll submit a PR when it is finished. I expect it to be much faster than the current one.

**Comment by jkinable** [comment]
Alright, the last iteration of performance improvements to Push-Relabel contributed by @AlexandruValeanu should allow us to permanently bury this issue. Thanks!
